lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -9020,6 +9020,196 @@ class DecomposeAtenTopkOp : public OpRewritePattern<AtenTopkOp> {
+                             Value &transposed) {
+      Type transposedType;
+      if (failed(getTransposedType(cast<BaseTensorType>(input.getType()), dimA,
+                                   dimB, transposedType)))
+        return failure();
+      Value cstDimA =
+          rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(dimA));
+      Value cstDimB =
+          rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(dimB));
+      transposed = rewriter.create<AtenTransposeIntOp>(loc, transposedType,
+                                                       input, cstDimA, cstDimB);
+      return success();
+    };
+
+    // Transpose if FFT dimension is not the last one
+    if (needTranspose) {
+      if (failed(transposeValue(rewriter, loc, self, dim, lastDim, self))) {
+        return failure();
+      }
+    }
+
+    // lhs = unsqueeze(self, -2) : (D x 1 x fftLength), D = [D_1, D_2, ...]
+    Value unsqueezeDim =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(-2));
+    auto unsqueezed = unsqueezeTensor(rewriter, op, self, unsqueezeDim);
+    if (failed(unsqueezed))
+      return rewriter.notifyMatchFailure(op,
+                                         "cannot generate unsqueezed tensor");
+    Value lhs = *unsqueezed;
+    Type dtype = inputType.getOptionalDtype();
+
+    Value real, complex;
+
+    for (const bool isRealPart : {true, false}) {
+
+      // coeff : (fftLength x outputFftDim)
+      ValueTensorType matrixType = ValueTensorType::get(
+          op.getContext(), SmallVector<int64_t>{fftLength, outputFftDim},
+          dtype);
+      Value coeffMatrix = getDFTMatmulCoeff(rewriter, loc, matrixType,
+                                            /*isRealPart=*/isRealPart);
+
+      // X = matmul(lhs, coeff) : (D x 1 x outputFftDim)
+      Value matmulRes = createBatchMatmul(rewriter, loc, lhs, coeffMatrix);
+
+      // Y = squeeze(X, -2) : (D x outputFftDim)
+      auto squeezed = squeezeTensor(rewriter, op, loc, -2, matmulRes);
+      if (failed(squeezed))
+        return rewriter.notifyMatchFailure(op,
+                                           "cannot generate squeezed tensor");
+
+      if (isRealPart) {
+        real = *squeezed;
+      } else {
+        complex = *squeezed;
+      }
+    }
+
+    // Pack components into a complex tensor
+    BaseTensorType realType = cast<BaseTensorType>(real.getType());
+    SmallVector<int64_t> stackSizes(realType.getSizes());
+    stackSizes.push_back(2);
+    Value sequence = rewriter.create<PrimListConstructOp>(
+        loc, ListType::get(op.getContext(), realType),
+        ValueRange{real, complex});
+    Type stackType = realType.getWithSizesAndDtype(stackSizes, dtype);
+    Value cstMinusOne =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(-1));
+    Value stack =
+        rewriter.create<AtenStackOp>(loc, stackType, sequence, cstMinusOne);
```

Comment:
I erroneously assumed that through optimization passes it would have been transformed to the optimal computation that you described, but indeed it doesn't. Also I'm not sure how an optimizing transformation for this case should be expressed. For simplicity I'll change the decomposition to the form that you suggest, although, by doing so, the decomposition becomes slightly less readable.